### PR TITLE
vindexes: return unknown params

### PR DESCRIPTION
**Pull request by maxenglander**
## Description

This PR adds an opt-in `vindexes.ParamValidating` interface that vindexes can implement in order to report unknown params passed to them during creation. Outside of unit tests, nothing is done with these unknown params in this PR.

Some other changes in this PR:

 * Validate tests.
 * Use a generalized method of testing vindex creation that can be used across vindex tests.
 * Unexports all forms of vindex creation except for `vindexes.CreateVindex`. Strictly speaking this isn't necessary for this PR, is meant to eliminate multiple code paths that do the same thing.

Why not just report unknown params via the `error` returned from `CreateVindex`? My concern here is that there may be Vitess users passing unknown params to Vindex creation and everything works perfectly fine for those users. Don't want to create a nasty surprise for them, so think it's better to roll this out in an opt-in fashion initially, and give the community a bigger heads up if we ever decide to error on unknown params.

### Use cases

In some cases when the VSchema contains parameters which are not understood by VTGate, VTGate will happily accept the VSchema without emitting any metrics or logs indicating that the VSchema contains unknown fields. 

This is probably fine in most cases for most people. But one situation where it could be an issue if the user meant for the unknown param (e.g. due to typo) to influence performance, where initially everything seems fine but later becomes a problem when load increases and performance degrades because the unknown param is not having the intended effect.

It would be great if we exposed unknown params as logs or metrics, with the option of opting into stricter error checking if desired.

### Next steps

Planned follow-up PRs:

 - Expose warnings in logs and metrics.
 - When `--vschema-strict-mode=error`, turn warnings into errors.
 - Enrich `vindexes.Param` to include additional attributes like `Required()`, `Type()`, `DefaultValue()`, `AllowedValues()`, etc. Use those attributes to do additional input validation.

## Related Issue(s)

Addresses #12476 

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation ~was added or~ is not required

### CODE REVIEW

File: go/vt/vtgate/vindexes/binary.go [hunk 1]
```diff
@@ -36,9 +36,9 @@ type Binary struct {
 	name string
 }
 
-// NewBinary creates a new Binary.
-func NewBinary(name string, _ map[string]string) (Vindex, error) {
-	return &Binary{name: name}, nil
+// newBinary creates a new Binary.
+func newBinary(name string, _ map[string]string) (Vindex, []VindexWarning, error) {
```
Comment (maxenglander) [inline go/vt/vtgate/vindexes/binary.go:40]: Unexport in order to force vindex creation to go through `CreateVindex`.

File: go/vt/vtgate/vindexes/cfc_test.go [hunk 2]
```diff
@@ -30,94 +30,117 @@ import (
...
+		expectCost:         1,
+		expectErr:          expectErr,
+		expectIsUnique:     true,
+		expectNeedsVCursor: false,
+		expectString:       "cfc",
+		expectWarnings:     expectWarnings,
 	}
 }
 
-func TestCFCBuildCFC(t *testing.T) {
```
Comment (maxenglander) [inline go/vt/vtgate/vindexes/cfc_test.go:44]: Renamed and test cases consolidated into `TestCFCCreateVindex`. All other indexes have a similarly named test case `Test[IndexType]CreateVindex`.

File: go/vt/vtgate/vindexes/cfc_test.go [hunk 3]
```diff
@@ -30,94 +30,117 @@ import (
...
-				assert.Equal(t, 1, cfc.Cost())
-				assert.Equal(t, true, cfc.IsUnique())
-				assert.Equal(t, false, cfc.NeedsVCursor())
-			}
-		})
-	}
+	testCreateVindexes(t, cases)
+}
+
+func TestCFCCreateVindexOptions(t *testing.T) {
```
Comment (maxenglander) [inline go/vt/vtgate/vindexes/cfc_test.go:128]: This test case was extracted from `TestCFCBuildCFC`. It was the only case that didn't inspected internal struct params, and didn't fit the generalized `testCreateVindexes` framework.

File: go/vt/vtgate/vindexes/consistent_lookup_test.go [hunk 4]
```diff
@@ -55,22 +109,6 @@ func TestConsistentLookupInit(t *testing.T) {
 	}
 }
 
-func TestConsistentLookupInfo(t *testing.T) {
```
Comment (maxenglander) [inline go/vt/vtgate/vindexes/consistent_lookup_test.go:58]: `Test[VindexType]Info()` tests are now handled in `Test[VindexType]CreateVindex` tests, are redundant and are removed in this PR.

File: go/vt/vtgate/vindexes/binary_test.go [hunk 5]
```diff
@@ -24,25 +24,72 @@ import (
...
 	"vitess.io/vitess/go/vt/key"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
 )
 
 var binOnlyVindex SingleColumn
 
 func init() {
-	vindex, _ := CreateVindex("binary", "binary_varchar", nil)
+	vindex, warnings, err := CreateVindex("binary", "binary_varchar", nil)
```
Comment (maxenglander) [inline go/vt/vtgate/vindexes/binary_test.go:38]: Adding warnings validations anywhere `CreateVindex` is used in tests in the `vindexes` package.

**Review by harshit-gangal — CHANGES_REQUESTED** [review]
The idea behind this PR is good.

This needs an RFC of its own as this is a big change in how vindexes are registered and defined.

I can see a potential issue with backward incompatibility here as all the existing vindex definition needs to be updated. Also, if `AllowUnknownParams` is set to `false` the Vschema creation will become version dependent.

How those will be tackled would be good to be written down in an RFC.

**Comment by maxenglander** [comment]
> This needs an RFC of its own as this is a big change in how vindexes are registered and defined.

Sounds good I can do that next.

> Also, if AllowUnknownParams is set to false the Vschema creation will become version dependent.

In this PR, `AllowUnknownParams` only returns warnings, not errors, so it should cause any backwards compatibility issues. My planned next PR was to emit these warnings in logs and metrics, which would also be backwards compatible.

**Comment by maxenglander** [comment]
RFC submitted. This PR corresponds with "option 1". https://github.com/vitessio/vitess/issues/13041

**Comment by maxenglander** [comment]
> so it should cause any backwards compatibility issues.

I now understand that you probably meant backwards incompatible issues for out-of-tree Vindex plugins. Agree that is unfortunate but I think it would be a relatively easy change for out-of-tree plugin maintainers.

**Comment by maxenglander** [comment]
I updated this to do the first part of option 1 in https://github.com/vitessio/vitess/issues/13041: store the params as `map[string]string` directly on vindexes, and report warnings when `InvalidParamErrors()` is called. 

The remaining parts (extract errors and store them on `KeyspaceSchema`, emit metrics) plan to do in separate PRs.

File: go/vt/vtgate/vindexes/main_test.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
+package vindexes
```
Comment (harshit-gangal) [inline go/vt/vtgate/vindexes/main_test.go:1]: need license header

File: go/vt/vtgate/vindexes/region_json.go [hunk 7]
```diff
@@ -32,12 +32,17 @@ import (
 	"vitess.io/vitess/go/vt/log"
 )
 
+const (
+	regionJSONParamRegionBytes = "region_bytes"
+	regionJSONParamRegionMap   = "region_map"
+)
```
Comment (harshit-gangal) [inline go/vt/vtgate/vindexes/region_json.go:38]: this is not referred in `rb, err := strconv.Atoi(m["region_bytes"])`

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/vindexes/region_json.go (hunk 7 above)
Comment (maxenglander) [reply go/vt/vtgate/vindexes/region_json.go:38]: whoops fixed

**Review by harshit-gangal — APPROVED** [review]

**Comment by harshit-gangal** [comment]
@maxenglander a website doc PR would be needed for this change.

**Comment by maxenglander** [comment]
Hey @harshit-gangal what would I document in the website? This PR doesn't expose any user-facing controls, and AFAIK the website doesn't document the `vindexes.CreateVindex` function nor the APIs for implementing a custom Vindex.

File: go/vt/vtgate/vindexes/binarymd5.go [hunk 8]
```diff
@@ -26,18 +26,23 @@ import (
...
 }
 
-// NewBinaryMD5 creates a new BinaryMD5.
-func NewBinaryMD5(name string, _ map[string]string) (Vindex, error) {
-	return &BinaryMD5{name: name}, nil
+// newBinaryMD5 creates a new BinaryMD5.
+func newBinaryMD5(name string, params map[string]string) (Vindex, error) {
+	return &BinaryMD5{
+		name:   name,
+		params: params,
```
Comment (GuptaManan100) [inline go/vt/vtgate/vindexes/binarymd5.go:44]: Why not store the unknown params directly like it is done for `binary` vindex?

**Review by GuptaManan100 — APPROVED** [review]

**Comment by harshit-gangal** [comment]
> I see that the interface `ParamValidating` is implemented by all the vindexes but the method `UnknownParams` is not called anywhere except in the tests.
> 
> The PR description talks about the flag but I could not find that flag and it using the `UnknownParams` method.

I see it is planned as a follow-up PR. In that case, we can update the website doc for param validation in that over here `https://vitess.io/docs/18.0/reference/features/vindexes/`

File: go/vt/vtgate/vindexes/binarymd5.go (hunk 8 above)
Comment (maxenglander) [reply go/vt/vtgate/vindexes/binarymd5.go:44]: Hey @GuptaManan100 the `binary_md5` vindex does not recognize any params, so any params that are passed in are unknown. So in this case it makes no difference from either a functional perspective or in terms of how memory gets used. But I see the argument for doing things the same way in all cases. I'll plan on making this more consistent in follow-up PR.

Comment (GuptaManan100) [reply go/vt/vtgate/vindexes/binarymd5.go:44]: Okay 👍

**Pull request merged**